### PR TITLE
Refactoring to prepare for FormFactorWrapper containing particles in multiple layers

### PR DESCRIPTION
**Pull request by waltervh**

### CODE REVIEW

File: Core/Multilayer/SSCApproximationStrategy.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ double SSCApproximationStrategy1::evaluateForList(const SimulationElement& sim_e
         return 0.0;
     for (size_t i = 0; i < m_formfactor_wrappers.size(); ++i) {
         complex_t ff = m_precomputed_ff1[i];
-        double fraction = m_formfactor_wrappers[i]->m_abundance / m_total_abundance;
+        double fraction = m_formfactor_wrappers[i]->getRelativeAbundance() / m_total_abundance;
         diffuse_intensity += fraction * std::norm(ff);
```
Comment (jwuttke) [inline Core/Multilayer/SSCApproximationStrategy.cpp]: really relative / total ?

File: Core/Multilayer/SSCApproximationStrategy.cpp [hunk 2]
```diff
@@ -120,7 +120,7 @@ double SSCApproximationStrategy2::evaluateForList(const SimulationElement& sim_e
         return 0.0;
     for (size_t i = 0; i < m_formfactor_wrappers.size(); ++i) {
         Eigen::Matrix2cd ff = m_precomputed_ff2[i];
-        double fraction = m_formfactor_wrappers[i]->m_abundance / m_total_abundance;
+        double fraction = m_formfactor_wrappers[i]->getRelativeAbundance() / m_total_abundance;
         diffuse_matrix += fraction * (ff * sim_element.getPolarization() * ff.adjoint());
```
Comment (jwuttke) [inline Core/Multilayer/SSCApproximationStrategy.cpp]: ditto

File: Core/Aggregate/FormFactorWrapper.cpp [hunk 3]
```diff
@@ -24,6 +24,16 @@ FormFactorWrapper* FormFactorWrapper::clone() const
...
+{
+    return mp_ff;
+}
+
+const IFormFactor *FormFactorWrapper::formfactor() const
+{
+    return mp_ff;
+}
+
 void FormFactorWrapper::scaleRelativeAbundance(double total_abundance)
```
Comment (jwuttke) [inline Core/Aggregate/FormFactorWrapper.cpp]: makes me unhappy. I'd so much prefer such trivial one-liners in the h file.

File: Core/Aggregate/FormFactorWrapper.cpp [hunk 4]
```diff
@@ -15,10 +15,21 @@
...
+
+void FormFactorWrapper::scaleRelativeAbundance(double total_abundance)
+{
+    if (total_abundance>0.0) {
+        m_abundance /= total_abundance;
+        return;
+    }
+    throw Exceptions::LogicErrorException("FormFactorWrapper::scaleRelativeAbundance: "
+                                              "Trying to scale with non strictly positive factor.");
+}
```
Comment (waltervh) [inline Core/Aggregate/FormFactorWrapper.cpp]: I prefer to see first the normal behavior in a method so I immediately see what it does under normal circumstances.

File: Core/Aggregate/FormFactorWrapper.cpp (hunk 3 above)
Comment (waltervh) [inline Core/Aggregate/FormFactorWrapper.cpp]: These methods (formfactor()) were temporary constructions that are now replaced with forwarding methods. They have been removed in my current branch completely.

File: Core/Aggregate/FormFactorWrapper.cpp (hunk 4 above)
Comment (jwuttke) [inline Core/Aggregate/FormFactorWrapper.cpp]: Good topic for next group discussion. What if several checks are needed?

    if (!A) {
        if (!B) {
            default_action
        } else
            throw "B!";
    } else
        throw "!A"

**Comment by jwuttke** [comment]
I'll merge, but we need to come back to exception handling.

**Pull request merged**